### PR TITLE
Support complex in eigh and svd

### PR DESCRIPTION
**Pull request by infrub**
This PR solves https://github.com/cupy/cupy/issues/180 and https://github.com/cupy/cupy/issues/1476 .
Now it is able to calculate svd decomp of complex matrix and to calculate eigen of hermitian matrix.

### CODE REVIEW

**Comment by unnonouno** [comment]
It looks good. Please fix test as well.

**Comment by toslunar** [comment]
- [x] svd test
- [ ] eigh test

Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
@unnonouno Could you review the PR?

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 1]
```diff
@@ -70,8 +73,15 @@ def test_mode(self):
...
 class TestSVD(unittest.TestCase):
 
+    def setUp(self):
+        global seed_testsvd
+        self.seed = seed_testsvd
+        seed_testsvd += 1
+
     @testing.for_float_dtypes(no_float16=True)
-    def check_usv(self, array, dtype):
+    def check_usv(self, shape, dtype):
```
Comment (unnonouno) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: Can you test with complex dtypes?

**Comment by themightyoarfish** [comment]
Could this be moved along?

**Comment by themightyoarfish** [comment]
Googling around I found a thread from the Nvidia Dev forums where the answer was basically that it's not competitive with mkl. It seems magma has a fast gpu implementation though, but I guess there's no hope for making that a dependency.

File: tests/cupy_tests/linalg_tests/test_decomposition.py (hunk 1 above)
Comment (toslunar) [reply tests/cupy_tests/linalg_tests/test_decomposition.py]: `for_float_dtypes` tests complex dtypes by default.

Comment (unnonouno) [reply tests/cupy_tests/linalg_tests/test_decomposition.py]: wow, thank you

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
LGTM

**Comment by unnonouno** [comment]
@themightyoarfish Currently we don't have a plan to support magma in CuPy.

**Comment by themightyoarfish** [comment]
Yes of course. If I wanted to get the magma SVD running just for myself, could you give some pointers on where to start? Like which parts of Cupy I would have to touch?

**Pull request merged**